File: gajeong/Programmers/LV2/광물캐기.js
function solution(picks, minerals) {
  var answer = 0;
  const res = [];
  const N = minerals.length;
  const DFS = (idx, picks, tired, now) => {
    if (now !== -1) {
      for (i = 0; i < 5; i++) {
        if (idx >= N) break;
        if (now == 0) {
          tired += 1;
        } else if (now == 1) {
          if (minerals[idx] === 'diamond') tired += 5;
          else tired += 1;
        } else {
          if (minerals[idx] === 'diamond') tired += 25;
          else if (minerals[idx] === 'iron') tired += 5;
          else tired += 1;
        }
        ++idx;
      }
    }

    //종료조건
    if (picks.every((el) => el === 0) || idx >= N) return tired;
    //진행   광물 ->
    for (let i = 0; i < 3; i++) {
      if (picks[i] > 0) {
        picks[i] -= 1;
        DFS(idx, picks, tired, i) === undefined
          ? ''
          : res.push(DFS(idx, picks, tired, i));

        picks[i] += 1;
      }
    }
  };

  DFS(0, picks, 0, -1);
  console.log(Math.min(...res));
  return answer;
}

solution(
  [1, 3, 2],
  ['diamond', 'diamond', 'diamond', 'iron', 'iron', 'diamond', 'iron', 'stone']
);
solution(
  [0, 1, 1],
  [
    'diamond',
    'diamond',
    'diamond',
    'diamond',
    'diamond',
    'iron',
    'iron',
    'iron',
    'iron',
    'iron',
    'diamond',
  ]
);

File: jiwoo/Programmers/광물 캐기.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> picks_copy;
vector<string> minerals_copy;

int ans = 987654321;

void dfs(int idx, int ret)
{
    if (minerals_copy.size() == idx || (picks_copy[0] == 0 && picks_copy[1] == 0 && picks_copy[2] == 0))
    {
        ans = min(ans, ret);
        return;
    }

    for (int i = 0; i < 3; i++)
    {
        if (picks_copy[i] != 0)
        {
            picks_copy[i]--;
            int tmp = 0;
            for (int j = 0; j < 5; j++)
            {
                if (idx + j >= minerals_copy.size())
                {
                    ans = min(ans, ret + tmp);
                    picks_copy[i]++;
                    return;
                }
                if (i == 0)
                {
                    if (minerals_copy[idx + j] == "diamond")
                    {
                        tmp += 1;
                    }
                    else if (minerals_copy[idx + j] == "iron")
                    {
                        tmp += 1;
                    }
                    else if (minerals_copy[idx + j] == "stone")
                    {
                        tmp += 1;
                    }
                }
                else if (i == 1)
                {
                    if (minerals_copy[idx + j] == "diamond")
                    {
                        tmp += 5;
                    }
                    else if (minerals_copy[idx + j] == "iron")
                    {
                        tmp += 1;
                    }
                    else if (minerals_copy[idx + j] == "stone")
                    {
                        tmp += 1;
                    }
                }
                else if (i == 2)
                {
                    if (minerals_copy[idx + j] == "diamond")
                    {
                        tmp += 25;
                    }
                    else if (minerals_copy[idx + j] == "iron")
                    {
                        tmp += 5;
                    }
                    else if (minerals_copy[idx + j] == "stone")
                    {
                        tmp += 1;
                    }
                }
            }
            dfs(idx + 5, ret + tmp);
            picks_copy[i]++;
        }
    }
}

int solution(vector<int> picks, vector<string> minerals)
{
    int answer = 0;
    
    picks_copy.assign(picks.begin(), picks.end());
    minerals_copy.assign(minerals.begin(), minerals.end());

    dfs(0, 0);

    answer = ans;

    return answer;
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}